### PR TITLE
Implement order-independent purge

### LINKED ISSUE #1152: Error in rgbasm.5 symbol interpolation example
Trying to assemble the first example in the 'Symbol interpolation' section causes an error.

The error is on the last line (`PURGE topic, meaning, {meaning}`) -- `Interpolated symbol "meaning" does not exist`

I'm assuming the symbols/arguments are purged left to right and interpolated after the preceding one has been purged. So, in effect, it's doing `PURGE meaning` _then_ `PURGE {meaning}`.

Swapping the order of the arguments `PURGE topic, {meaning}, meaning` eliminates the error and purges the all of the defined symbols (i.e. `topic, answer, meaning`).

### PR DESCRIPTION
**Pull request by quinnyo**
This changes how purge works:
- `sym_Purge` collects the symbols in a list instead of removing them
- `sym_FlushPurged` removes/frees all of the symbols at once (it gets called after the purge list is processed)

I haven't written C for quite a long time, so if something looks like an obvious mistake, it surely is.

This fixes #1152. It implements the solution described in the [comment](https://github.com/gbdev/rgbds/issues/1152#issuecomment-1719016000) on that issue.

I wasn't sure about adding a test case -- is it as simple as something like:

```
; test/asm/purge-deferred.asm

DEF prefix EQUS "cool"
DEF {prefix}banana EQU 1

PURGE prefix, {prefix}banana

ASSERT !DEF(prefix)
ASSERT !DEF(coolbanana)
```

### CODE REVIEW

**Comment by Rangi42** [comment]
Yes, that test case fails currently and would pass here, so it works perfectly.

File: src/asm/symbol.c [hunk 1]
```diff
@@ -276,6 +276,13 @@ static bool isReferenced(struct Symbol const *sym)
 	return sym->ID != (uint32_t)-1;
 }
 
+struct SymbolListNode {
+	struct Symbol *this;
```
Comment (Rangi42) [inline src/asm/symbol.c:280]: :pick: C++ is close enough to C that it feels weird to use `this` as a variable. (Same goes for `class`, etc.)

File: src/asm/symbol.c [hunk 2]
```diff
@@ -288,15 +295,47 @@ void sym_Purge(char const *symName)
...
-		// free(sym->macro) because the expansion may be purging itself.
-		hash_RemoveElement(symbols, sym->name);
-		// TODO: ideally, also unref the file stack nodes
-		free(sym);
+		struct SymbolListNode *end = purgeList;
+
+		while (end && end->next)
+			end = end->next;
+
+		struct SymbolListNode *node = malloc(sizeof(struct SymbolListNode));
```
Comment (Rangi42) [inline src/asm/symbol.c:303]:
```suggestion
		struct SymbolListNode *node = malloc(sizeof(*node));
```

File: src/asm/symbol.c [hunk 3]
```diff
@@ -288,15 +295,47 @@ void sym_Purge(char const *symName)
...
-
-		// FIXME: this leaks sym->macro for SYM_EQUS and SYM_MACRO, but this can't
-		// free(sym->macro) because the expansion may be purging itself.
-		hash_RemoveElement(symbols, sym->name);
-		// TODO: ideally, also unref the file stack nodes
-		free(sym);
+		struct SymbolListNode *end = purgeList;
+
+		while (end && end->next)
+			end = end->next;
```
Comment (Rangi42) [inline src/asm/symbol.c:301]: :wrench: Seeking to the end of the purge list each time could be avoided if you prepend each symbol to the beginning, instead of appending to the end.

File: src/asm/symbol.c (hunk 1 above)
Comment (aaaaaa123456789) [reply src/asm/symbol.c:280]: Only because your highlighter is reminding you of it :P
I use C++ keywords as C identifiers all the time (`new` is a common one) and I've never seen any issues come out of that, even when building with C++-capable compilers (GCC, Clang, etc.). The toolchains don't seem to be confused.

File: src/asm/symbol.c [hunk 4]
```diff
@@ -288,15 +295,47 @@ void sym_Purge(char const *symName)
...
+			end->next = node;
+		else
+			purgeList = node;
+	}
+}
+
+// Remove purged symbols
+void sym_FlushPurged(void)
+{
+	struct SymbolListNode *node = purgeList;
```
Comment (Rangi42) [inline src/asm/symbol.c:318]: :wrench: You don't need this variable; just work with `purgeList` directly.

File: src/asm/symbol.c [hunk 5]
```diff
@@ -288,15 +295,47 @@ void sym_Purge(char const *symName)
...
+
+	purgeList = NULL;
+	while (node) {
+		struct SymbolListNode *next = node->next;
+		struct Symbol *sym = node->this;
+
+		free(node);
+		node = next;
```
Comment (Rangi42) [inline src/asm/symbol.c:326]:
```suggestion
	while (purgeList) {
		struct Symbol *sym = purgeList->this;
		struct SymbolListNode *next = purgeList->next;

		free(purgeList);
		purgeList = next;
```

File: src/asm/symbol.c [hunk 6]
```diff
@@ -288,15 +295,47 @@ void sym_Purge(char const *symName)
...
+
+		struct SymbolListNode *node = malloc(sizeof(struct SymbolListNode));
+
+		node->this = sym;
+		node->next = NULL;
+
+		if (end)
+			end->next = node;
+		else
+			purgeList = node;
```
Comment (Rangi42) [inline src/asm/symbol.c:311]:
```suggestion
		struct SymbolListNode *node = malloc(sizeof(*node));

		node->this = sym;
		node->next = purgeList;
		purgeList = node;
```

File: src/asm/symbol.c [hunk 7]
```diff
@@ -288,15 +295,47 @@ void sym_Purge(char const *symName)
...
+
+	purgeList = NULL;
+	while (node) {
+		struct SymbolListNode *next = node->next;
+		struct Symbol *sym = node->this;
+
+		free(node);
+		node = next;
+
+		if (sym) {
```
Comment (Rangi42) [inline src/asm/symbol.c:328]: Will there ever be a case where `sym` can be `NULL` here? If not, don't have an `if`, just `assert(sym)`.

File: src/asm/symbol.c (hunk 1 above)
Comment (Rangi42) [reply src/asm/symbol.c:280]: I didn't say "it could break", I said "it feels weird". :P And we *were* considering to convert things gradually to C++, although the Rust rewrite probably superseded that.

**Review by Rangi42 — CHANGES_REQUESTED** [review]
Thanks for taking initiative to fix this! Just some style+efficiency edits to make.

File: src/asm/symbol.c (hunk 1 above)
Comment (quinnyo) [reply src/asm/symbol.c:280]: I felt a little bit weird doing it, but I also felt free! (will use `sym`, thanks)

**Comment by quinnyo** [comment]
Suggestions look good, thanks.
Do you prefer squashed commits? (I usually would in a case like this)

**Comment by quinnyo** [comment]
All of your suggestions made sense (and made my code better!) so they got used verbatim and I also added the test case. Let me know if I need to do anything else!

**Review by Rangi42 — APPROVED** [review]

File: test/asm/purge-deferred.asm [hunk 8]
```diff
@@ -0,0 +1,10 @@
+DEF prefix EQUS "cool"
```
Comment (ISSOtm) [inline test/asm/purge-deferred.asm:1]: Ideally, a comment would indicate somewhere *what* the test is checking for. (This becomes a life-saver when some tests break from unrelated changes, like syntax breaks or bugfixes.)

File: src/asm/symbol.c [hunk 9]
```diff
@@ -276,6 +276,13 @@ static bool isReferenced(struct Symbol const *sym)
 	return sym->ID != (uint32_t)-1;
 }
 
+struct SymbolListNode {
+	struct Symbol *sym;
+	struct SymbolListNode *next;
+};
+
+static struct SymbolListNode *purgeList = NULL;
```
Comment (ISSOtm) [inline src/asm/symbol.c:284]: I prefer avoiding global variables, as they are Not Great for various reasons. (Primarily, they require a lot more work when trying to turn something unique into several instances thereof.)

Unfortunately, the way to fix this is to make the `purge_list` in the grammar an array of symbols[^cxx], and delete each of them at the end of the line.
Why unfortunately? Because that requires making deeper changes to the grammar file—none difficult, thankfully, but if you're not used to it, it'd require some learning...

I'm willing to provide guidance/mentoring to avoid you having to read the entire GNU Bison manual, if you want to.

[^cxx]: \*grumbles\* *God*, give me C++ `std::vector` (or `Vec` lol) any time over our 13th manually-implemented specialisation of this goddamn data type...

File: src/asm/symbol.c [hunk 10]
```diff
@@ -276,6 +276,13 @@ static bool isReferenced(struct Symbol const *sym)
 	return sym->ID != (uint32_t)-1;
 }
 
+struct SymbolListNode {
+	struct Symbol *sym;
+	struct SymbolListNode *next;
```
Comment (ISSOtm) [inline src/asm/symbol.c:281]: I know RGBASM has a lot of those, mostly because it was the path of least resistance and we haven't fixed them, but I'd prefer at least all newly-introduced code to use arrays instead of linked lists.

Linked lists require more allocations, and are generally slower to traverse; they're good at insertion and deletion of items at arbitrary points (and not much else), but we're not doing either of those here.

**Review by ISSOtm — CHANGES_REQUESTED** [review]
Oh, hey! Thank you for your interest in contributing! Feel free to reach us on https://gbdev.io/chat if you want to get more real-time feedback, as well.

File: src/asm/symbol.c (hunk 9 above)
Comment (Rangi42) [reply src/asm/symbol.c:284]: Oh neat, GFM has [footnotes](https://github.blog/changelog/2021-09-30-footnotes-now-supported-in-markdown-fields/).

**Review by ISSOtm — APPROVED** [review]
Well, thank you, Rangi, for addressing my concerns... but I think it would have been better to wait a little more to give @quinnyo a chance to reply, so he could get his feet wet if he wanted to.

I'll merge this regardless, as it wouldn't make sense to dismiss a good patch; but, @quinnyo, please know that my offer of mentoring / guidance still holds for any other PR you may choose to tackle after this one.

**Comment by Rangi42** [comment]
Ah, sorry about that.

**Comment by quinnyo** [comment]
Lucky I checked the PR before spending too much more time on this, I guess... Cheers
